File: cordova-plugin-outline/android/java/org/outline/shadowsocks/ShadowsocksConnectivity.java
// Copyright 2018 The Outline Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.outline.shadowsocks;

import java.io.BufferedReader;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.ConnectException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.nio.ByteBuffer;
import java.util.Locale;
import java.util.logging.Level;
import java.util.logging.Logger;
import java.util.Random;

/**
 * Class that performs connectivity tests on remote Shadowsocks servers.
 *
 */
public class ShadowsocksConnectivity {
  private static final Logger LOG = Logger.getLogger(ShadowsocksConnectivity.class.getName());
  private static final int SOCKS_RESPONSE_NUM_BYTES = 10;
  private static final byte SOCKS_CMD_CONNECT = 0x1;
  private static final byte SOCKS_ATYP_IPV4 = 1;
  private static final byte SOCKS_ATYP_DOMAINNAME = 3;;
  private static final byte SOCKS_VERSION = 0x5;
  private static final int SOCKS_HEADER_ATYP = 3;
  private static final int SOCKS_METHODS_HEADER_NUM_BYTES = 3;
  private static final int SOCKS_METHODS_RESPONSE_NUM_BYTES = 2;
  private static final byte SOCKS_N_METHODS = 1;
  private static final byte SOCKS_METHOD_NOAUTH = 0;
  private static final int SOCKS_TCP_HEADER_NUM_BYTES = 6;
  // We have chosen these domains due to their neutrality.
  private static final String[] CREDENTIALS_VALIDATION_DOMAINS = {"eff.org", "ietf.org", "w3.org",
                                                                  "wikipedia.org", "example.com"};
  private static final short CREDENTIALS_VALIDATION_PORT = 80;
  private static final int TCP_SOCKET_TIMEOUT_MS = 10000;
  private static final int UDP_SOCKET_TIMEOUT_MS = 1000;
  private static final int UDP_MAX_RETRY_ATTEMPTS = 5;
  private static final int UDP_MAX_BUFFER_NUM_BYTES = 512;
  private static final String DNS_RESOLVER_IP = "208.67.222.222";  // OpenDNS
  private static final short DNS_RESOLVER_PORT = 53;
  private static final int SERVER_CONNECT_MAX_ATTEMPTS = 3;
  private static final int SERVER_CONNECT_RETRY_SLEEP_MS = 500;

  // Returns whether the server is reachable at the supplied IP address.
  public static boolean isServerReachable(final String ip, int port) {
    return isAddressReachable(new InetSocketAddress(ip, port));
  }

  /**
   * Verifies that the remote server credentials are valid. Performs an end-to-end authentication
   * test by issuing an HTTP HEAD request to a target domain.
   */
  public static boolean validateServerCredentials(final String localProxyIp, final int localProxyPort) {
    LOG.fine("Starting server creds. validation.");
    Socket socket = null;
    DataOutputStream outputStream = null;
    DataInputStream inputStream = null;
    try {
      InetSocketAddress localProxyAddress = new InetSocketAddress(localProxyIp, localProxyPort);
      if (!waitForLocalShadowsocksServer(localProxyAddress)) {
        return false;
      }
      socket = new Socket();
      socket.setSoTimeout(TCP_SOCKET_TIMEOUT_MS);
      socket.connect(localProxyAddress);
      outputStream = new DataOutputStream(socket.getOutputStream());
      inputStream = new DataInputStream(socket.getInputStream());

      byte[] buffer = getSocksMethodsRequest();
      outputStream.write(buffer);
      buffer = new byte[SOCKS_METHODS_RESPONSE_NUM_BYTES];
      inputStream.readFully(buffer);  // Don't parse, we already know the response is NOAUTH

      final String targetDomain = chooseRandomDomain();
      buffer = getSocksTcpRequest(targetDomain);
      outputStream.write(buffer);
      buffer = new byte[SOCKS_RESPONSE_NUM_BYTES];
      inputStream.readFully(buffer);  // Don't parse, this is a fake response

      final String httpRequest =
          String.format(Locale.ROOT, "HEAD / HTTP/1.1\r\nHost: %s\r\n\r\n", targetDomain);
      outputStream.write(httpRequest.getBytes());
      BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
      final String httpResponse = reader.readLine();
      return httpResponse != null && httpResponse.startsWith("HTTP/1.1");
    } catch (IOException e) {
      LOG.log(Level.WARNING, "Got exception in server creds. validation", e);
    } finally {
      closeSocket(socket);
    }
    return false;
  }

  /**
   * Verifies that the server has enabled UDP forwarding. Sends a DNS request through the remote on
   * the specified local proxy address.
   */
  public static boolean isUdpForwardingEnabled(final String localProxyIp, final int localProxyPort) {
    LOG.fine("Starting UDP forwarding validation");
    DatagramSocket socket = null;
    try {
      InetSocketAddress localProxyAddress = new InetSocketAddress(localProxyIp, localProxyPort);
      if (!waitForLocalShadowsocksServer(localProxyAddress)) {
        return false;
      }
      byte[] buffer = getSocksUdpRequest();
      DatagramPacket dnsRequest = new DatagramPacket(buffer, buffer.length, localProxyAddress);
      DatagramPacket dnsResponse = new DatagramPacket(new byte[UDP_MAX_BUFFER_NUM_BYTES],
                                                      UDP_MAX_BUFFER_NUM_BYTES);
      socket = new DatagramSocket();
      socket.setSoTimeout(UDP_SOCKET_TIMEOUT_MS);

      int attempt = 1;
      boolean success = false;
      while (attempt <= UDP_MAX_RETRY_ATTEMPTS) {
        LOG.info(String.format(Locale.ROOT, "Checking remote UDP forwarding (%d/%d)", attempt,
            UDP_MAX_RETRY_ATTEMPTS));
        socket.send(dnsRequest);
        try {
          socket.receive(dnsResponse);
          success = true;
          break;
        } catch (SocketTimeoutException e) {
          LOG.warning("UDP forwarding validation timed out.");
          ++attempt;
        }
      }
      return success;
    } catch (IOException e) {
      LOG.log(Level.SEVERE, "Unexpected exception in UDP forwarding validation", e);
      return false;
    } finally {
      if (socket != null) {
        socket.close();
      }
    }
  }

  // Returns whether |address| is reachable.
  private static boolean isAddressReachable(InetSocketAddress address) {
    Socket socket = null;
    try {
      socket = new Socket();
      socket.connect(address, TCP_SOCKET_TIMEOUT_MS);
      return true;
    } catch (Exception e) {
      LOG.log(Level.WARNING, "Connection failure while determining address reachability", e);
    } finally {
      closeSocket(socket);
    }
    return false;
  }

  // Returns whether |serverAddress| is reachable. Attempts to connect up to three times,
  // sleeping in between.
  private static boolean waitForLocalShadowsocksServer(InetSocketAddress serverAddress) {
    for (int attempt = 0; attempt < SERVER_CONNECT_MAX_ATTEMPTS; ++attempt) {
      if (isAddressReachable(serverAddress)) {
        return true;
      }
      try {
        Thread.sleep(SERVER_CONNECT_RETRY_SLEEP_MS);
      } catch (InterruptedException e) {
        // We shouldn't get here, but if we do return true to avoid a false negative.
        LOG.warning("Connectivity test interrupted");
        return true;
      }
    }
    LOG.severe(String.format(Locale.ROOT, "Failed to reach server after %d attempts, giving up",
        SERVER_CONNECT_MAX_ATTEMPTS));
    return false;
  }

  // Synthesizes a SOCKS UDP request with a DNS query as payload.
  private static byte[] getSocksUdpRequest() throws IOException {
    InetAddress dnsResolverAddress = null;
    try {
      dnsResolverAddress = InetAddress.getByName(DNS_RESOLVER_IP);
    } catch (UnknownHostException e) {
      throw new IOException("Failed to compose a SOCKS UDP request", e);
    }
    ByteBuffer buffer = ByteBuffer.allocate(UDP_MAX_BUFFER_NUM_BYTES);
    // Construct SOCKS UDP header
    buffer.position(SOCKS_HEADER_ATYP);
    buffer.put(SOCKS_ATYP_IPV4).put(dnsResolverAddress.getAddress()).putShort(DNS_RESOLVER_PORT)
    // Copy DNS Request
        .put(getDnsRequest());
    return buffer.array();
  }

  // Synthesizes a DNS request for google.com
  private static byte[] getDnsRequest() throws IOException {
    return new byte[] {
      0, 0,  // [0-1]   query ID
      1, 0,  // [2-3]   flags; byte[2] = 1 for recursion desired (RD).
      0, 1,  // [4-5]   QDCOUNT (number of queries)
      0, 0,  // [6-7]   ANCOUNT (number of answers)
      0, 0,  // [8-9]   NSCOUNT (number of name server records)
      0, 0,  // [10-11] ARCOUNT (number of additional records)
      6, 'g', 'o', 'o', 'g', 'l', 'e',
      3, 'c', 'o', 'm',
      0,  // null terminator of FQDN (root TLD)
      0, 1, // QTYPE, set to A
      0, 1  // QCLASS, set to 1 = IN (Internet)
    };
  }

  // Synthesizes a SOCKS TCP CONNECT header to |domain| on port 80.
  private static byte[] getSocksTcpRequest(final String domain) {
    ByteBuffer buffer = ByteBuffer.allocate(SOCKS_TCP_HEADER_NUM_BYTES +
                                            domain.length() + 1 /* domain length byte */);
    buffer.put(SOCKS_VERSION).put(SOCKS_CMD_CONNECT);
    buffer.position(SOCKS_HEADER_ATYP);
    buffer.put(SOCKS_ATYP_DOMAINNAME).put((byte)domain.length()).put(domain.getBytes())
        .putShort(CREDENTIALS_VALIDATION_PORT);
    return buffer.array();
  }

  private static byte[] getSocksMethodsRequest() {
    ByteBuffer buffer = ByteBuffer.allocate(SOCKS_METHODS_HEADER_NUM_BYTES);
    return buffer.put(SOCKS_VERSION).put(SOCKS_N_METHODS).put(SOCKS_METHOD_NOAUTH).array();
  }

  private static void closeSocket(Socket socket) {
    try {
      if (socket != null && !socket.isClosed()) {
        socket.close();
      }
    } catch (IOException e) { /* Ignore */ }
  }

  // Returns a random domain from |CREDENTIALS_VALIDATION_DOMAINS|.
  private static final String chooseRandomDomain() {
    int index = new Random().nextInt(CREDENTIALS_VALIDATION_DOMAINS.length);
    return CREDENTIALS_VALIDATION_DOMAINS[index];
  }
}

File: cordova-plugin-outline/apple/vpn/ShadowsocksConnectivity.m
// Copyright 2018 The Outline Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#import "ShadowsocksConnectivity.h"
#include <arpa/inet.h>
#if TARGET_OS_IPHONE
#import <Shadowsocks_iOS/shadowsocks.h>
#else
#import <Shadowsocks_macOS/shadowsocks.h>
#endif

@import CocoaAsyncSocket;
@import CocoaLumberjack;

static char *const kShadowsocksLocalAddress = "127.0.0.1";

static char *const kDnsResolverAddress = "208.67.222.222";  // OpenDNS
static const uint16_t kDnsResolverPort = 53;
static const size_t kDnsRequestNumBytes = 28;

static const size_t kSocksHeaderNumBytes = 10;
static const uint8_t kSocksMethodsResponseNumBytes = 2;
static const size_t kSocksConnectResponseNumBytes = 10;
static const uint8_t kSocksVersion = 0x5;
static const uint8_t kSocksMethodNoAuth = 0x0;
static const uint8_t kSocksCmdConnect = 0x1;
static const uint8_t kSocksAtypIpv4 = 0x1;
static const uint8_t kSocksAtypDomainname = 0x3;

static const NSTimeInterval kTcpSocketTimeoutSecs = 10.0;
static const NSTimeInterval kUdpSocketTimeoutSecs = 1.0;
static const long kSocketTagHttpRequest = 100;
static const int kUdpForwardingMaxChecks = 5;
static const uint16_t kHttpPort = 80;

@interface ShadowsocksConnectivity ()<GCDAsyncSocketDelegate, GCDAsyncUdpSocketDelegate>

@property(nonatomic) uint16_t shadowsocksPort;

@property(nonatomic, copy) void (^udpForwardingCompletion)(BOOL);
@property(nonatomic, copy) void (^reachabilityCompletion)(BOOL);
@property(nonatomic, copy) void (^credentialsCompletion)(BOOL);

@property(nonatomic) dispatch_queue_t dispatchQueue;
@property(nonatomic) GCDAsyncUdpSocket *udpSocket;
@property(nonatomic) GCDAsyncSocket *credentialsSocket;
@property(nonatomic) GCDAsyncSocket *reachabilitySocket;

@property(nonatomic) bool isRemoteUdpForwardingEnabled;
@property(nonatomic) bool areServerCredentialsValid;
@property(nonatomic) bool isServerReachable;
@property(nonatomic) int udpForwardingNumChecks;
@end

@implementation ShadowsocksConnectivity

- (id)initWithPort:(uint16_t)shadowsocksPort {
  self = [super init];
  if (self) {
    _shadowsocksPort = shadowsocksPort;
    _dispatchQueue = dispatch_queue_create("ShadowsocksConnectivity", DISPATCH_QUEUE_SERIAL);
  }
  return self;
}

#pragma mark - UDP Forwarding

struct socks_udp_header {
  uint16_t rsv;
  uint8_t frag;
  uint8_t atyp;
  uint32_t addr;
  uint16_t port;
};

- (void)isUdpForwardingEnabled:(void (^)(BOOL))completion {
  DDLogInfo(@"Starting remote UDP forwarding check.");
  self.isRemoteUdpForwardingEnabled = false;
  self.udpForwardingNumChecks = 0;
  self.udpForwardingCompletion = completion;
  self.udpSocket =
      [[GCDAsyncUdpSocket alloc] initWithDelegate:self delegateQueue:self.dispatchQueue];
  struct in_addr dnsResolverAddress;
  if (!inet_aton(kDnsResolverAddress, &dnsResolverAddress)) {
    DDLogError(@"Failed to convert DNS resolver IP.");
    [self udpForwardingCheckDone:false];
    return;
  }
  struct socks_udp_header socksHeader = {
      .atyp = kSocksAtypIpv4,
      .addr = dnsResolverAddress.s_addr,  // Already in network order
      .port = htons(kDnsResolverPort)};
  uint8_t *dnsRequest = [self getDnsRequest];
  size_t packetNumBytes = kSocksHeaderNumBytes + kDnsRequestNumBytes;
  uint8_t socksPacket[packetNumBytes];
  memset(socksPacket, 0, packetNumBytes);
  memcpy(socksPacket, &socksHeader, kSocksHeaderNumBytes);
  memcpy(socksPacket + kSocksHeaderNumBytes, dnsRequest, kDnsRequestNumBytes);

  NSData *packetData = [[NSData alloc] initWithBytes:socksPacket length:packetNumBytes];

  dispatch_source_t timer =
      dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, self.dispatchQueue);
  if (!timer) {
    DDLogError(@"Failed to create timer");
    [self udpForwardingCheckDone:false];
    return;
  }
  dispatch_source_set_timer(timer, dispatch_time(DISPATCH_TIME_NOW, 0),
                            kUdpSocketTimeoutSecs * NSEC_PER_SEC, 0);
  __weak ShadowsocksConnectivity *weakSelf = self;
  dispatch_source_set_event_handler(timer, ^{
    if (++weakSelf.udpForwardingNumChecks > kUdpForwardingMaxChecks ||
        weakSelf.isRemoteUdpForwardingEnabled) {
      dispatch_source_cancel(timer);
      if (!weakSelf.isRemoteUdpForwardingEnabled) {
        [weakSelf udpForwardingCheckDone:false];
      }
      [weakSelf.udpSocket close];
      return;
    }
    DDLogDebug(@"Checking remote server's UDP forwarding (%d of %d).",
               weakSelf.udpForwardingNumChecks, kUdpForwardingMaxChecks);
    [weakSelf.udpSocket sendData:packetData
                          toHost:[[NSString alloc] initWithUTF8String:kShadowsocksLocalAddress]
                            port:self.shadowsocksPort
                     withTimeout:kUdpSocketTimeoutSecs
                             tag:0];
    if (![weakSelf.udpSocket receiveOnce:nil]) {
      DDLogError(@"UDP socket failed to receive data");
    }
  });
  dispatch_resume(timer);
}

// Returns a byte representation of a DNS request for "google.com".
- (uint8_t *)getDnsRequest {
  static uint8_t kDnsRequest[] = {
      0, 0,  // [0-1]   query ID
      1, 0,  // [2-3]   flags; byte[2] = 1 for recursion desired (RD).
      0, 1,  // [4-5]   QDCOUNT (number of queries)
      0, 0,  // [6-7]   ANCOUNT (number of answers)
      0, 0,  // [8-9]   NSCOUNT (number of name server records)
      0, 0,  // [10-11] ARCOUNT (number of additional records)
      6, 'g', 'o', 'o', 'g', 'l', 'e', 3, 'c', 'o', 'm',
      0,     // null terminator of FQDN (root TLD)
      0, 1,  // QTYPE, set to A
      0, 1   // QCLASS, set to 1 = IN (Internet)
  };
  return kDnsRequest;
}

#pragma mark - GCDAsyncUdpSocketDelegate

- (void)udpSocket:(GCDAsyncUdpSocket *)sock
    didNotSendDataWithTag:(long)tag
               dueToError:(NSError *)error {
  DDLogError(@"Failed to send data on UDP socket");
}

- (void)udpSocket:(GCDAsyncUdpSocket *)sock
       didReceiveData:(NSData *)data
          fromAddress:(NSData *)address
    withFilterContext:(id)filterContext {
  if (!self.isRemoteUdpForwardingEnabled) {
    // Only report success if it hasn't been done so already.
    [self udpForwardingCheckDone:true];
  }
}

- (void)udpForwardingCheckDone:(BOOL)enabled {
  DDLogInfo(@"Remote UDP forwarding %@", enabled ? @"enabled" : @"disabled");
  self.isRemoteUdpForwardingEnabled = enabled;
  if (self.udpForwardingCompletion != NULL) {
    self.udpForwardingCompletion(self.isRemoteUdpForwardingEnabled);
    self.udpForwardingCompletion = NULL;
  }
}

#pragma mark - Credentials

struct socks_methods_request {
  uint8_t ver;
  uint8_t nmethods;
  uint8_t method;
};

struct socks_request_header {
  uint8_t ver;
  uint8_t cmd;
  uint8_t rsv;
  uint8_t atyp;
};

- (void)checkServerCredentials:(void (^)(BOOL))completion {
  DDLogInfo(@"Starting server creds. validation");
  self.areServerCredentialsValid = false;
  self.credentialsCompletion = completion;
  self.credentialsSocket =
      [[GCDAsyncSocket alloc] initWithDelegate:self delegateQueue:self.dispatchQueue];
  NSError *error;
  [self.credentialsSocket
      connectToHost:[[NSString alloc] initWithUTF8String:kShadowsocksLocalAddress]
             onPort:self.shadowsocksPort
        withTimeout:kTcpSocketTimeoutSecs
              error:&error];
  if (error) {
    DDLogError(@"Unable to connect to local Shadowsocks server.");
    [self serverCredentialsCheckDone];
    return;
  }

  struct socks_methods_request methodsRequest = {
      .ver = kSocksVersion, .nmethods = 0x1, .method = kSocksMethodNoAuth};
  NSData *methodsRequestData =
      [[NSData alloc] initWithBytes:&methodsRequest length:sizeof(struct socks_methods_request)];
  [self.credentialsSocket writeData:methodsRequestData withTimeout:kTcpSocketTimeoutSecs tag:0];
  [self.credentialsSocket readDataToLength:kSocksMethodsResponseNumBytes
                               withTimeout:kTcpSocketTimeoutSecs
                                       tag:0];

  size_t socksRequestHeaderNumBytes = sizeof(struct socks_request_header);
  NSString *domain = [self chooseRandomDomain];
  uint8_t domainNameNumBytes = domain.length;
  size_t socksRequestNumBytes = socksRequestHeaderNumBytes + domainNameNumBytes +
                                sizeof(uint16_t) /* port */ +
                                sizeof(uint8_t) /* domain name length */;

  struct socks_request_header socksRequestHeader = {
      .ver = kSocksVersion, .cmd = kSocksCmdConnect, .atyp = kSocksAtypDomainname};
  uint8_t socksRequest[socksRequestNumBytes];
  memset(socksRequest, 0x0, socksRequestNumBytes);
  memcpy(socksRequest, &socksRequestHeader, socksRequestHeaderNumBytes);
  socksRequest[socksRequestHeaderNumBytes] = domainNameNumBytes;
  memcpy(socksRequest + socksRequestHeaderNumBytes + sizeof(uint8_t), [domain UTF8String],
         domainNameNumBytes);
  uint16_t httpPort = htons(kHttpPort);
  memcpy(socksRequest + socksRequestHeaderNumBytes + sizeof(uint8_t) + domainNameNumBytes,
         &httpPort, sizeof(uint16_t));

  NSData *socksRequestData =
      [[NSData alloc] initWithBytes:socksRequest length:socksRequestNumBytes];
  [self.credentialsSocket writeData:socksRequestData withTimeout:kTcpSocketTimeoutSecs tag:0];
  [self.credentialsSocket readDataToLength:kSocksConnectResponseNumBytes
                               withTimeout:kTcpSocketTimeoutSecs
                                       tag:0];

  NSString *httpRequest =
      [[NSString alloc] initWithFormat:@"HEAD / HTTP/1.1\r\nHost: %@\r\n\r\n", domain];
  [self.credentialsSocket
        writeData:[NSData dataWithBytes:[httpRequest UTF8String] length:httpRequest.length]
      withTimeout:kTcpSocketTimeoutSecs
              tag:kSocketTagHttpRequest];
  [self.credentialsSocket readDataWithTimeout:kTcpSocketTimeoutSecs tag:kSocketTagHttpRequest];
  [self.credentialsSocket disconnectAfterReading];
}

// Returns a statically defined array containing domain names for validating server credentials.
+ (const NSArray *)getCredentialsValidationDomains {
  static const NSArray *kCredentialsValidationDomains;
  static dispatch_once_t kDispatchOnceToken;
  dispatch_once(&kDispatchOnceToken, ^{
    // We have chosen these domains due to their neutrality.
    kCredentialsValidationDomains =
        @[ @"eff.org", @"ietf.org", @"w3.org", @"wikipedia.org", @"example.com" ];
  });
  return kCredentialsValidationDomains;
}

// Returns a random domain from |kCredentialsValidationDomains|.
- (NSString *)chooseRandomDomain {
  const NSArray *domains = [ShadowsocksConnectivity getCredentialsValidationDomains];
  int index = arc4random_uniform((uint32_t)domains.count);
  return domains[index];
}

// Calls |credentialsCompletion| once with |areServerCredentialsValid|.
- (void)serverCredentialsCheckDone {
  DDLogInfo(@"Server creds. %@.", self.areServerCredentialsValid ? @"succeeded" : @"failed");
  if (self.credentialsCompletion != NULL) {
    self.credentialsCompletion(self.areServerCredentialsValid);
    self.credentialsCompletion = NULL;
  }
}

#pragma mark - Reachability

- (void)isReachable:(NSString *)host port:(uint16_t)port completion:(void (^)(BOOL))completion {
  DDLogInfo(@"Starting server reachability check.");
  self.isServerReachable = false;
  self.reachabilityCompletion = completion;
  self.reachabilitySocket =
      [[GCDAsyncSocket alloc] initWithDelegate:self delegateQueue:self.dispatchQueue];
  NSError *error;
  [self.reachabilitySocket connectToHost:host
                                  onPort:port
                             withTimeout:kTcpSocketTimeoutSecs
                                   error:&error];
  if (error) {
    DDLogError(@"Unable to connect to Shadowsocks server.");
    return;
  }
}

// Calls |reachabilityCompletion| once with |isServerReachable|.
- (void)reachabilityCheckDone {
  DDLogInfo(@"Server %@.", self.isServerReachable ? @"reachable" : @"unreachable");
  if (self.reachabilityCompletion != NULL) {
    self.reachabilityCompletion(self.isServerReachable);
    self.reachabilityCompletion = NULL;
  }
}

#pragma mark - GCDAsyncSocketDelegate

- (void)socket:(GCDAsyncSocket *)sock didReadData:(NSData *)data withTag:(long)tag {
  // We don't need to inspect any of the data, as the SOCKS responses are hardcoded in ss-local and
  // the fact that we have read the HTTP response indicates that the server credentials are valid.
  if (tag == kSocketTagHttpRequest && data != nil) {
    NSString *httpResponse = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    self.areServerCredentialsValid = httpResponse != nil && [httpResponse hasPrefix:@"HTTP/1.1"];
  }
}

- (void)socket:(GCDAsyncSocket *)sock didConnectToHost:(NSString *)host port:(uint16_t)port {
  if ([self.reachabilitySocket isEqual:sock]) {
    self.isServerReachable = true;
    [self.reachabilitySocket disconnect];
  }
}

- (void)socketDidDisconnect:(GCDAsyncSocket *)sock withError:(NSError *)error {
  if ([self.reachabilitySocket isEqual:sock]) {
    [self reachabilityCheckDone];
  } else {
    [self serverCredentialsCheckDone];
  }
}

@end
